docs/src/modules/utils/useQueryParameterState.ts

```diff
@@ -0,0 +1,77 @@
+import * as React from 'react';
+import debounce from 'lodash/debounce';
+import { useRouter } from 'next/router';
+
+/**
+ * Similar to `React.useState`, but it syncs back the current state to a query
+ * parameter in the url, therefore it only supports strings. Wrap the result with
+ * parse/stringify logic if more complex values are needed.
+ *
+ * REMARK: this doesn't listen for router changes (yet) to update back the state.
+ */
+export default function useQueryParameterState(
+  name: string,
+  initialValue = '',
+): [string, (newValue: string) => void] {
+  const initialValueRef = React.useRef(initialValue);
+
+  const router = useRouter();
+
+  const queryParamValue = router.query[name];
+
+  const urlValue = Array.isArray(queryParamValue) ? queryParamValue[0] : queryParamValue;
+
+  const [state, setState] = React.useState(urlValue || initialValue);
+
+  const setUrlValue = React.useMemo(
+    () =>
+      debounce((newValue = '') => {
```

Comment:
lodash debounce has a `debounced.cancel` method (ours is `clear`). We could consider calling it to avoid URL search leaks when unmounting this component.